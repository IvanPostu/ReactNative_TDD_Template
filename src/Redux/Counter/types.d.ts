import {Action} from 'redux';

export interface CounterStateType {
  counter: number;
}

export enum counterActionTypeConstants {
  INCREMENT = '@Counter/INCREMENT',
  DECREMENT = '@Counter/DECREMENT',
  SET_ZERO = '@Counter/SET_ZERO',
  SET_VALUE = '@Counter/SET_VALUE',
}

export interface IncrementActionType extends Action {
  type: typeof counterActionTypeConstants.INCREMENT;
}

export interface DecrementActionType extends Action {
  type: typeof counterActionTypeConstants.DECREMENT;
}

export interface SetZeroActionType extends Action {
  type: typeof counterActionTypeConstants.SET_ZERO;
}

export interface SetValueActionType extends Action {
  type: typeof counterActionTypeConstants.SET_VALUE;
  payload: number;
}

export type CounterRootActionType =
  | IncrementActionType
  | DecrementActionType
  | SetZeroActionType
  | SetValueActionType;
