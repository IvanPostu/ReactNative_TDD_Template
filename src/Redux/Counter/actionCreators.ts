import * as T from './types.d';

export function incrementActionCreator(): T.IncrementActionType {
  return {
    type: T.counterActionTypeConstants.INCREMENT,
  };
}

export function decrementActionCreator(): T.DecrementActionType {
  return {
    type: T.counterActionTypeConstants.DECREMENT,
  };
}

export function setZeroActionCreator(): T.SetZeroActionType {
  return {
    type: T.counterActionTypeConstants.SET_ZERO,
  };
}

export function setValueActionCreator(val: number): T.SetValueActionType {
  return {
    type: T.counterActionTypeConstants.SET_VALUE,
    payload: val,
  };
}
