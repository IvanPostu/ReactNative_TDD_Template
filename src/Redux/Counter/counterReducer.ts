import {Reducer} from 'redux';

import {
  CounterRootActionType,
  CounterStateType,
  counterActionTypeConstants as T,
} from './types.d';

const initialState: CounterStateType = {
  counter: 0,
};

export const counterReducer: Reducer<
  CounterStateType,
  CounterRootActionType
> = (
  state: CounterStateType = initialState,
  action: CounterRootActionType,
): CounterStateType => {
  switch (action.type) {
    case T.INCREMENT:
      return {
        ...state,
        counter: state.counter + 1,
      };

    case T.DECREMENT:
      return {
        ...state,
        counter: state.counter - 1,
      };

    case T.SET_ZERO:
      return {
        ...state,
        counter: 0,
      };

    case T.SET_VALUE:
      return {
        ...state,
        counter: action.payload,
      };

    default:
      return state;
  }
};
