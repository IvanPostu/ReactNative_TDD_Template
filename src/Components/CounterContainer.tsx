import {ReduxGlobalStateType} from '@/Redux';
import React, {ReactElement, useState} from 'react';
import {Button, StyleSheet, Text, TextInput, View} from 'react-native';
import {connect} from 'react-redux';
import {bindActionCreators, Dispatch} from 'redux';
import * as actionCreators from '@/Redux/Counter/actionCreators';
import Button1 from './Button';

type CounterContainerPropTypes = {} & ReturnType<typeof mapStateToProps> &
  ReturnType<typeof mapDispatchToProps>;

export const CounterContainer = (
  props: CounterContainerPropTypes,
): ReactElement => {
  const [count, setCount] = useState(0);

  function onCounterTextChange(text: string) {
    try {
      const validNumber = text.replace(/[^0-9]/g, '');
      setCount(+validNumber);
    } catch {}
  }

  return (
    <View style={styles.root}>
      <Text testID="outputValue" style={styles.text}>
        {String(props.count)}
      </Text>
      <View>
        <Button
          testID="incrementBtn"
          onPress={() => props.incrementActionCreator()}
          title="Increment"
        />
        <Button
          testID="decrementBtn"
          onPress={() => props.decrementActionCreator()}
          title="Decriment"
        />
        <TextInput
          testID="inputValue"
          keyboardType="numeric"
          style={styles.input}
          value={String(count)}
          onChangeText={onCounterTextChange}
        />
        <Button
          testID="setValueButton"
          onPress={() => {
            props.setValueActionCreator(count);
          }}
          title="Set value"
        />
        <Button1
          isLoading={false}
          onPress={() => {
            console.log(1);
          }}
          title="qq"
          transparent={false}
        />
      </View>
    </View>
  );
};

const mapStateToProps = (state: ReduxGlobalStateType) => ({
  count: state.counterReducer.counter,
});

function mapDispatchToProps(dispatch: Dispatch) {
  return bindActionCreators(actionCreators, dispatch);
}

export default connect(mapStateToProps, mapDispatchToProps)(CounterContainer);

const styles = StyleSheet.create({
  root: {
    height: 400,
    backgroundColor: 'black',
  },
  text: {
    color: '#fff',
  },
  input: {
    backgroundColor: '#fff',
  },
});
