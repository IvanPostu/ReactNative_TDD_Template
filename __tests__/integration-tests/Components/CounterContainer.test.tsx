import 'react-native';
import {fireEvent, render} from '@testing-library/react-native';
import CounterContainer from '@/Components/CounterContainer';
import React from 'react';
import {expect, test, describe} from '@jest/globals';

import {combineReducers, createStore} from 'redux';
import {Provider} from 'react-redux';
import {counterReducer} from '@/Redux/Counter/counterReducer';

const rootReducer = combineReducers({
  counterReducer,
});
const store = createStore(rootReducer, {counterReducer: {counter: 10}});

describe('CounterContainer test', () => {
  test('CounterContainer set count value correctly', async () => {
    const {getByTestId} = render(
      <Provider store={store}>
        <CounterContainer />
      </Provider>,
    );
    const setValueButton = getByTestId('setValueButton');
    const inputValue = getByTestId('inputValue');
    const outputValue = getByTestId('outputValue');
    const incrementBtn = getByTestId('incrementBtn');
    const decrementBtn = getByTestId('decrementBtn');

    fireEvent.press(incrementBtn);
    fireEvent.press(incrementBtn);
    expect(outputValue.props.children).toEqual('12');

    fireEvent.press(decrementBtn);
    fireEvent.press(decrementBtn);
    fireEvent.press(decrementBtn);
    fireEvent.press(decrementBtn);
    expect(outputValue.props.children).toEqual('8');

    await fireEvent.changeText(inputValue, '1022');
    fireEvent.press(setValueButton);
    expect(outputValue.props.children).toEqual('1022');

    expect(store.getState().counterReducer.counter).toBe(1022);
  });
});
