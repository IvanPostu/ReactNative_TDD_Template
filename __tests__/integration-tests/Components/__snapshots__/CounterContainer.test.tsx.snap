// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CounterContainer component test Renders correctly (take snapshot) 1`] = `
<View
  style={
    Object {
      "backgroundColor": "black",
      "height": 400,
    }
  }
>
  <Text
    style={
      Object {
        "color": "#fff",
      }
    }
    testID="outputValue"
  >
    10
  </Text>
  <View>
    <View
      accessibilityRole="button"
      accessibilityState={Object {}}
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID="incrementBtn"
    >
      <View
        style={
          Array [
            Object {},
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#007AFF",
                "fontSize": 18,
                "margin": 8,
                "textAlign": "center",
              },
            ]
          }
        >
          Increment
        </Text>
      </View>
    </View>
    <View
      accessibilityRole="button"
      accessibilityState={Object {}}
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID="decrementBtn"
    >
      <View
        style={
          Array [
            Object {},
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#007AFF",
                "fontSize": 18,
                "margin": 8,
                "textAlign": "center",
              },
            ]
          }
        >
          Decriment
        </Text>
      </View>
    </View>
    <TextInput
      allowFontScaling={true}
      keyboardType="numeric"
      onChangeText={[Function]}
      rejectResponderTermination={true}
      style={
        Object {
          "backgroundColor": "#fff",
        }
      }
      testID="inputValue"
      underlineColorAndroid="transparent"
      value="0"
    />
    <View
      accessibilityRole="button"
      accessibilityState={Object {}}
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID="setValueButton"
    >
      <View
        style={
          Array [
            Object {},
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#007AFF",
                "fontSize": 18,
                "margin": 8,
                "textAlign": "center",
              },
            ]
          }
        >
          Set value
        </Text>
      </View>
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#74B3CE",
            "borderRadius": 4,
            "height": 50,
            "justifyContent": "center",
            "marginVertical": 10,
            "width": "100%",
          }
        }
      >
        <Text
          style={
            Object {
              "color": "white",
              "fontSize": 16,
            }
          }
        >
          qq
        </Text>
      </View>
    </View>
    <Image
      source={
        Object {
          "testUri": "../../../src/Assets/cat.png",
        }
      }
    />
  </View>
</View>
`;
